packages/compile-solidity/compileWithPragmaAnalysis.js

```diff
@@ -0,0 +1,123 @@
+  // releases are ordered from newest to oldest
+  return solcReleases.find(version => {
+    return semverExpressions.every(expression =>
+      semver.satisfies(version, expression)
+    );
+  });
+};
+
+const throwCompilerVersionNotFound = ({ path, semverExpressions }) => {
+  const message =
+    `Could not find a single version of the Solidity compiler that ` +
+    `satisfies the following semver expressions obtained from your source ` +
+    `files' pragma statements: ${semverExpressions.join(" - ")}. ` +
+    `${OS.EOL}Please check the pragma statements for ${path} and its imports.`;
+  throw new Error(message);
+};
+
+const compileWithPragmaAnalysis = async ({ paths, options }) => {
+  const supplierOptions = {
+    events: options.events,
+    solcConfig: options.compilers.solc
+  };
+  const compilerSupplier = new CompilerSupplier(supplierOptions);
+  const { releases } = await compilerSupplier.getReleases();
+
+  // collect sources by the version of the Solidity compiler that they require
+  const versionsAndSources = {};
+  for (const path of paths) {
+    const source = await fse.readFile(path, "utf8");
+
+    const parserVersion = findNewestSatisfyingVersion({
+      solcReleases: releases,
+      semverExpressions: [getSemverExpression(source)]
+    });
+    if (!parserVersion) {
+      throwCompilerVersionNotFound({
+        path,
+        semverExpressions: [getSemverExpression(source)]
+      });
+    }
+
+    // allSources is of the format { [filename]: string }
+    const { allSources } = await Profiler.requiredSourcesForSingleFile(
+      options.with({
+        path,
+        base_path: options.contracts_directory,
+        resolver: options.resolver,
+        compilers: {
+          solc: {
+            version: parserVersion
+          }
+        }
+      })
+    );
+
+    // get an array of all the semver expressions in the sources
+    const semverExpressions = await getSemverExpressions(
+      Object.values(allSources)
+    );
+    const newestSatisfyingVersion = findNewestSatisfyingVersion({
+      solcReleases: releases,
+      semverExpressions
+    });
+    if (!newestSatisfyingVersion) {
+      throwCompilerVersionNotFound({
+        path,
+        semverExpressions
+      });
+    }
+
+    if (versionsAndSources[newestSatisfyingVersion]) {
+      versionsAndSources[newestSatisfyingVersion] = {
+        ...versionsAndSources[newestSatisfyingVersion],
+        ...allSources
+      };
+    } else {
+      versionsAndSources[newestSatisfyingVersion] = allSources;
+    }
+  }
+
+  const compilations = [];
+  for (const compilerVersion of Object.keys(versionsAndSources)) {
+    const compilationOptions = options.merge({
+      compilers: {
+        solc: {
+          version: compilerVersion
+        }
+      }
+    });
+
+    const compilation = await run(
+      versionsAndSources[compilerVersion],
+      normalizeOptions(compilationOptions)
```

Comment:
I think the original `options` was already normalized, and `compilationOptions` is tightly controlled by us. I'm thinking we don't need to normalize again here.

Reply:
These options have not been normalized yet in this flow and it seems like the main point of this normalization is backwards compatibility? However we are guaranteed that it will be a `Config` object so you are right that it is probably unnecessary! I'll go ahead and remove this. There is another call to this method in `index.js` that can probably also be removed.

Reply:
:+1: Looks like `options.with` normalizes everything already.